core/components/molecules/form/field/field-input.js

```diff
@@ -0,0 +1,122 @@
+import React from 'react'
+import ReactDOM from 'react-dom'
+
+/*
+  There can be just one or multiple input fields inside a Form.Field.
+
+  We want to pass our accessibility props to the first input element.
+  This might be the only child in the field as well.
+
+  However, multiple fields would typically be wrapped inside one
+  layout component like StackLayout which means a single child for us.
+  The layout element should never get the id.
+
+  Here's what we can do:
+
+  1. In render
+
+     If the first child is a cosmos component that renders an input element
+     without a container like ['TextInput', 'TextArea', 'Select'], we can
+     attach an id rightaway.
+
+     If it's one of the cosmos components that has a more complex nested structure
+     like ['Switch', 'Radio', 'Checkbox'] or if the first child is a layout component,
+     we can't attach an id on the parent, we handle this as a progressive enhancement
+     after render.
+
+  2. After render:
+
+     On componentDidMount, we find the first input element and set the id attribute
+     if it wasn't already set.
+
+     This means the rendered output is initially not accessible, but is made
+     accessible progressively.
+*/
+
+const cosmosInputWhitelist = ['TextInput', 'TextArea', 'Select']
+
+class FieldInput extends React.Component {
+  constructor(props) {
+    super(props)
+    this.firstChildRef = React.createRef()
+  }
+  render() {
+    const { Component, ...fieldProps } = this.props
+    /*
+        old API support, we proxy through all the props
+        to the input element
+      */
+    if (Component) return <Component {...fieldProps} />
+
+    /*
+      New API
+      Here be dragons. Read the story above to justify the following complexity
+    */
+    let { children, id } = fieldProps
+
+    /* find the first child */
+    let modifiedChildren = React.Children.toArray(children)
+    const firstChild = modifiedChildren[0]
+
+    /*
+      If the first child is one of the input elements in the whitelist,
+      we can make it accessible by attaching the id.
+
+      Otherwise we perform a progressive enhancement and attach an id
+      after render = componentDidMount.
+
+      We keep track of the first child by attaching a ref to it.
+    */
+
+    let firstChildProps = {}
+
+    const componentName = firstChild.type.displayName
+    if (cosmosInputWhitelist.includes(componentName)) firstChildProps.id = id
+
+    /*
+      If it's a cosmos component, it uses styled-components,
+      which means we need to use innerRef
+
+      Note: This will give an error for a non cosmos element :/
+    */
+    firstChildProps.innerRef = this.firstChildRef
```

Comment:
The fact that this gives an error if you don't use a styled-component is stupid. We might be able to come up with a fix for it. 

I didn't invest time in this because this is [solved](https://www.styled-components.com/docs/api#deprecated-innerref-prop) in styled-component@v4